sys/senml/senml.c

```diff
@@ -0,0 +1,357 @@
+    }
+
+    pack->num = (array_num + 1);
+
+    return 0;
+}
+
+
+int senml_encode_json_s(const senml_pack_t *pack, char *output, size_t len)
+{
+    if (len < 3) {
+        DEBUG("ERROR: cannot encode pack to JSON because output buffer is too small\n");
+        return -1;
+    }
+
+    // using len - 3 for content to leave space for closing } and ] and '\0'
+    ssize_t cnt_len = len - 3;
+
+    size_t insert_pos = 0;
+
+    output[0] = '[';
+    insert_pos++;
+    cnt_len--;
+
+    int chars_written;
+
+    if (!pack->base_info) {
+        goto encode_records;
+    }
+
+    output[insert_pos] = '{';
+    insert_pos++;
+    cnt_len--;
+
+    if (pack->base_info->version) {
+        chars_written = snprintf(&output[insert_pos], cnt_len, "\"bver\":%u,",
+                                 pack->base_info->version);
+
+        if (chars_written >= cnt_len || chars_written < 0) {
+            return -1;
+        }
+
+        insert_pos += chars_written;
+        cnt_len -= chars_written;
+    }
+
+    if (pack->base_info->base_name) {
+        chars_written = snprintf(&output[insert_pos], cnt_len, "\"bn\":\"%s\",",
+                                 pack->base_info->base_name);
+
+        if (chars_written >= cnt_len || chars_written < 0) {
+            return -1;
+        }
+
+        insert_pos += chars_written;
+        cnt_len -= chars_written;
+    }
+
+    if (pack->base_info->base_time != 0) {
+        chars_written = snprintf(&output[insert_pos], cnt_len, "\"bt\":%f,",
+                                 pack->base_info->base_time);
+
+        if (chars_written >= cnt_len || chars_written < 0) {
+            return -1;
+        }
+
+        insert_pos += chars_written;
+        cnt_len -= chars_written;
+    }
+
+    if (pack->base_info->base_unit) {
+        chars_written = snprintf(&output[insert_pos], cnt_len, "\"bu\":\"%s\",",
+                                 pack->base_info->base_unit);
+
+        if (chars_written >= cnt_len || chars_written < 0) {
+            return -1;
+        }
+
+        insert_pos += chars_written;
+        cnt_len -= chars_written;
+    }
+
+    if (pack->base_info->base_value != 0) {
+        chars_written = snprintf(&output[insert_pos], cnt_len, "\"bv\":%f,",
+                                 pack->base_info->base_value);
+
+        if (chars_written >= cnt_len || chars_written < 0) {
+            return -1;
+        }
+
+        insert_pos += chars_written;
+        cnt_len -= chars_written;
+    }
+
+    encode_records:
+
+    if (insert_pos == 1) {
+        output[insert_pos] = '{';
+        insert_pos++;
+        cnt_len--;
+    }
+
+    for (size_t i = 0; i < pack->num; i++) {
+        senml_record_t *curr_record = &(pack->records[i]);
+
+        if (i != 0) {
+            output[insert_pos] = '{';
+            insert_pos++;
+            cnt_len--;
+        }
+
+        if (curr_record->name) {
+            chars_written = snprintf(&output[insert_pos], cnt_len, "\"n\":\"%s\",",
```

Comment:
`cnt_len` may be zero at the end of the loop, so it will be decremented to `-1` in the next iteration after inserting `{`.

Abstract execution trace of `cnt_len`:

```C
if (curr_record->name) {
  chars_written = snprintf(&output[insert_pos], cnt_len, "\"n\":\"%s\",",
                           curr_record->name);

  if (chars_written >= cnt_len || chars_written < 0) {
    return -1;
  }

  insert_pos += chars_written;
  cnt_len -= chars_written;

  //////////////////
  // cnt_len >= 1 //
  //////////////////
 }

// ...

// decrease insert_pos to overwrite trailing comma
insert_pos--;
cnt_len++;
//////////////////
// cnt_len >= 2 //
//////////////////

output[insert_pos] = '}';
insert_pos++;
cnt_len--;
//////////////////
// cnt_len >= 1 //
//////////////////

output[insert_pos] = ',';
insert_pos++;
cnt_len--;
//////////////////
// cnt_len >= 0 //
//////////////////

/////////////////////
// next iteratrion //
/////////////////////

i++;

senml_record_t *curr_record = &(pack->records[i]);

if (i != 0) {
  output[insert_pos] = '{';
  insert_pos++;
  cnt_len--;
 }

///////////////////
// cnt_len >= -1 //
///////////////////

if (curr_record->name) {
    /////////////////////////////////
    // cnt_len is casted to size_t //
    //         UNDERFLOW!          //
    /////////////////////////////////
    chars_written = snprintf(&output[insert_pos], cnt_len, "\"n\":\"%s\",",
                             curr_record->name); 

/// ...

```